Code/GraphMol/Substruct/SubstructMatch.h

```diff
@@ -23,81 +23,108 @@ class MolBundle;
 
 //! \brief used to return matches from substructure searching,
 //!   The format is (queryAtomIdx, molAtomIdx)
-typedef std::vector<std::pair<int, int> > MatchVectType;
+typedef std::vector<std::pair<int, int>> MatchVectType;
+
+struct RDKIT_SUBSTRUCTMATCH_EXPORT SubstructMatchParameters {
+  bool useChirality;  //!< Use chirality in determining whether or not
+                      //!< atoms/bonds match
+  bool aromaticMatchesConjugated;  //!< Aromatic and conjugated bonds match each
+                                   //!< other
+  bool useQueryQueryMatches;  //!< Consider query-query matches, not just simple
+                              //!< matches
+  bool recursionPossible;     //!< Allow recursive queries
+  bool uniquify;              //!< uniquify (by atom index) match results
+  unsigned int maxMatches;    //!< maximum number of matches to return
+  int numThreads;             //!< number of threads to use when multi-threading
+                              //!< is possible. 0 selects the number of
+                              //!< concurrent threads supported by the hardware
+                              //!< negative values are added to the number of
+                              //!< concurrent threads supported by the hardware
+
+  SubstructMatchParameters()
+      : useChirality(false),
+        aromaticMatchesConjugated(false),
+        useQueryQueryMatches(false),
+        recursionPossible(true),
+        uniquify(true),
+        maxMatches(1000),
+        numThreads(1){};
+};
 
 //! Find a substructure match for a query in a molecule
 /*!
-    \param mol       The ROMol to be searched
-    \param query     The query ROMol
-    \param matchVect Used to return the match
-                     (pre-existing contents will be deleted)
-    \param recursionPossible  flags whether or not recursive matches are allowed
-    \param useChirality  use atomic CIP codes as part of the comparison
-    \param useQueryQueryMatches  if set, the contents of atom and bond queries
-                                 will be used as part of the matching
+    \param mol         The ROMol to be searched
+    \param query       The query ROMol
+    \param matchParams Parameters controlling the matching
 
-    \return whether or not a match was found
+    \return The matches, if any
 
 */
-RDKIT_SUBSTRUCTMATCH_EXPORT bool SubstructMatch(const ROMol &mol, const ROMol &query,
-                    MatchVectType &matchVect, bool recursionPossible = true,
-                    bool useChirality = false,
-                    bool useQueryQueryMatches = false);
+RDKIT_SUBSTRUCTMATCH_EXPORT std::vector<MatchVectType> SubstructMatch(
+    const ROMol &mol, const ROMol &query,
+    const SubstructMatchParameters &params);
```

Comment:
If we had default SubstructMatchParameters this could be:

```
std::vector<MatchVectType> SubstructMatch(
  const ROMol&,
  const ROMol&,
 const SubstructMatchParameters &=SubstructMatchParameters()
);
```

so the default usage would be:

```
for (auto matches: SubstructMatch(mol, query)) {
} 
```

Which is pretty sweet.

Reply:
nice! On it.

Reply:
Those changes are in